bindings/cs/rl.net/RlLogger.cs

```diff
@@ -0,0 +1,337 @@
+        {
+            int result = LoggerLogCbInteraction(this.DangerousGetHandle(), contextJson, response.DangerousGetHandle(), apiStatus.ToNativeHandleOrNullptrDangerous());
+            return result == NativeMethods.SuccessStatus;
+        }
+
+        public void Log(string contextJson, DecisionResponse response)
+        {
+            using (ApiStatus apiStatus = new ApiStatus())
+            {
+                if (!this.TryLog(contextJson, response, apiStatus))
+                {
+                    throw new RLException(apiStatus);
+                }
+            }
+        }
+
+        public bool Log(string eventId, float outcome, ApiStatus apiStatus = null)
+        {
+            return this.TryLog(eventId, outcome, apiStatus);
+        }
+
+        public bool TryLog(string eventId, float outcome, ApiStatus apiStatus = null)
+        {
+            int result = LoggerLogOutcomeF(this.DangerousGetHandle(), eventId, outcome, apiStatus.ToNativeHandleOrNullptrDangerous());
+            return result == NativeMethods.SuccessStatus;
+        }
+
+        public void Log(string eventId, float outcome)
+        {
+            using (ApiStatus apiStatus = new ApiStatus())
+            {
+                if (!this.TryLog(eventId, outcome, apiStatus))
+                {
+                    throw new RLException(apiStatus);
+                }
+            }
+        }
+
+        public bool Log(string eventId, string outcome, ApiStatus apiStatus = null)
+        {
+            return this.TryLog(eventId, outcome, apiStatus);
+        }
+
+        public bool TryLog(string eventId, string outcome, ApiStatus apiStatus = null)
+        {
+            int result = LoggerLogOutcomeJson(this.DangerousGetHandle(), eventId, outcome, apiStatus.ToNativeHandleOrNullptrDangerous());
+            return result == NativeMethods.SuccessStatus;
+        }
+
+        public void Log(string eventId, string outcome)
+        {
+            using (ApiStatus apiStatus = new ApiStatus())
+            {
+                if (!this.TryLog(eventId, outcome, apiStatus))
+                {
+                    throw new RLException(apiStatus);
+                }
+            }
+        }
+
+        private event EventHandler<ApiStatus> BackgroundErrorInternal;
+
+        // This event is thread-safe, because we do not hook/unhook the event in user-schedulable code anymore.
+        public event EventHandler<ApiStatus> BackgroundError
+        {
+            add
+            {
+                this.BackgroundErrorInternal += value;
+            }
+            remove
+            {
+                this.BackgroundErrorInternal -= value;
+            }
+        }
+
+        private event EventHandler<TraceLogEventArgs> OnTraceLoggerEventInternal;
+
+        // TODO:
+        /// <remarks>
+        /// Add/remove here is not thread safe.
+        /// </remarks>
+        public event EventHandler<TraceLogEventArgs> TraceLoggerEvent
```

Comment:
Make this thread safe by using a sync block.

Reply:
Switched to RlLoggerThreadUnsafe/ThreadLogger implementations exactly as we have for LiveModel. If there are remaining issues that has to be fixed in both places, let's do it in separate PR.